app/src/main/java/cl/mario/covid/ui/main/MainActivity.kt

```diff
@@ -23,29 +19,40 @@ class MainActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
-
-        binding.button.setOnClickListener {
-            val c = Calendar.getInstance()
-            val year = c.get(Calendar.YEAR)
-            val month = c.get(Calendar.MONTH)
-            val day = c.get(Calendar.DATE).minus(1)
-            val datePickerDialog = DatePickerDialog(this, { view, year, month, day ->
-                val monthformat = month.plus(1).toString().padStart(2, '0')
-                val dayformat = day.toString().padStart(2, '0')
-                covidViewModel.getCovidResults("$year-$monthformat-$dayformat")
-            }, year, month, day)
-
-            datePickerDialog.datePicker.maxDate =
-                System.currentTimeMillis() - 24 * 60 * 60 * 1000 //sacandole el día vieja escuela jajaj sorry por esto
-            datePickerDialog.show()
-        }
-
     }
 
     override fun onStart() {
         super.onStart()
         binding.viewModel = covidViewModel
         binding.lifecycleOwner = this
+        initObserver()
+        initListener()
         covidViewModel.getCovidResults()
     }
+
+    private fun initObserver() {
+        covidViewModel.covidInfoStateLiveData.observe(this) {
+            if(it is State.Error ){
+                binding.errorView.loadError(it.message){
+                    covidViewModel.getCovidResults(covidViewModel.lastDateRequest)
+                }
+            }
+        }
+    }
+
+    private fun initListener(){
+        binding.apply {
+            button.setOnClickListener {
+                openCalendar()
+            }
+        }
+    }
+
+    private fun openCalendar(){
+        val calendar = CalendarManager(this)
+        calendar.show()
+        calendar.setOnSelectedListener {
+            covidViewModel.getCovidResults(dateToStringApi(it))
+        }
+    }
```

Comment:
I know you can reduce more this code 😄 
Maybe, you could to make an utils like:
```

private fun initListener() { 
            binding.button.setOnClickListener {
                openCalendar(context) {
                       covidViewModel.getCovidResults(dateToStringApi(it))
              }
            }
        }
```

Why? Because other developer can use this method without to know the steps necessary for completing configuration.